pkg/proxy/ipvs/proxier.go

```diff
@@ -808,6 +808,42 @@ func (proxier *Proxier) syncProxyRules() {
 	// activeBindAddrs represents ip address successfully bind to DefaultDummyDevice in this round of sync
 	activeBindAddrs := map[string]bool{}
 
+	hasNodePort := false
+	for _, svc := range proxier.serviceMap {
+		svcInfo, ok := svc.(*serviceInfo)
+		if ok && svcInfo.NodePort() != 0 {
+			hasNodePort = true
+			break
+		}
+	}
+
+	// Both nodeAddresses and nodeIPs can be reused for all nodePort services
+	// and only need to be computed if we have at least one nodePort service.
+	var (
+		// List of node addresses to listen on if a nodePort is set.
+		nodeAddresses sets.String
+		// List of node IP addresses to be used as IPVS services if nodePort is
+		// set.
+		nodeIPs []net.IP
+	)
+
+	if hasNodePort {
+		nodeAddresses, err = utilproxy.GetNodeAddresses(proxier.nodePortAddresses, proxier.networkInterfacer)
+		if err != nil {
+			klog.Errorf("Failed to get node ip address matching nodeport cidr: %v", err)
```

Comment:
why not to skip iterating nodeAddresses in this case ?

Reply:
This is what will happen in practice. When `GetNodeAddresses` returns an error it also returns a nil value for `nodeAddresses` and this will cause the iteration to be skipped.

Reply:
This isn't directly obviously though, I agree with tedyu it's easier to follow with an explicity skip or check. You can use `nodeAddresses.Len()` to do an explicity check before iterating node addresses

Reply:
Done.